In [43]:
import pandas as pd
import numpy as np
import os
from sklearn.preprocessing import StandardScaler
import warnings
warnings.filterwarnings('ignore')

In [36]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV

In [27]:
bank_preprocess = pd.read_csv(os.getcwd()+'//Bank_Cleaning_Complete.csv')
bank_preprocess.head()

Unnamed: 0  age        job  marital   education housing loan month  \
0           0   56  housemaid  married       basic      no   no   may   
1           1   57   services  married  highschool      no   no   may   
2           2   37   services  married  highschool     yes   no   may   
3           3   40      admin  married       basic      no   no   may   
4           4   56   services  married  highschool      no  yes   may   

  day_of_week  duration  campaign  previous     poutcome  emp.var.rate  \
0         mon       261         1         0  nonexistent           1.1   
1         mon       149         1         0  nonexistent           1.1   
2         mon       226         1         0  nonexistent           1.1   
3         mon       151         1         0  nonexistent           1.1   
4         mon       307         1         0  nonexistent           1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no

In [28]:
bank_preprocess.dtypes

Unnamed: 0          int64
age                 int64
job                object
marital            object
education          object
housing            object
loan               object
month              object
day_of_week        object
duration            int64
campaign            int64
previous            int64
poutcome           object
emp.var.rate      float64
cons.price.idx    float64
cons.conf.idx     float64
euribor3m         float64
nr.employed       float64
y                  object
dtype: object

In [29]:
category_features = ['job','marital','education','housing','loan','month','day_of_week','poutcome']
bank_preprocess[category_features].head()

job  marital   education housing loan month day_of_week     poutcome
0  housemaid  married       basic      no   no   may         mon  nonexistent
1   services  married  highschool      no   no   may         mon  nonexistent
2   services  married  highschool     yes   no   may         mon  nonexistent
3      admin  married       basic      no   no   may         mon  nonexistent
4   services  married  highschool      no  yes   may         mon  nonexistent

In [30]:
bank_preprocess_one_hot = pd.get_dummies(bank_preprocess, columns = category_features)
bank_preprocess_one_hot.drop(columns=['Unnamed: 0'],inplace=True)
bank_preprocess_one_hot = pd.get_dummies(bank_preprocess_one_hot, columns = ['y'],drop_first=True)
bank_preprocess_one_hot.head()

age  duration  campaign  previous  emp.var.rate  cons.price.idx  \
0   56       261         1         0           1.1          93.994   
1   57       149         1         0           1.1          93.994   
2   37       226         1         0           1.1          93.994   
3   40       151         1         0           1.1          93.994   
4   56       307         1         0           1.1          93.994   

   cons.conf.idx  euribor3m  nr.employed  job_admin  ...  month_sep  \
0          -36.4      4.857       5191.0          0  ...          0   
1          -36.4      4.857       5191.0          0  ...          0   
2          -36.4      4.857       5191.0          0  ...          0   
3          -36.4      4.857       5191.0          1  ...          0   
4          -36.4      4.857       5191.0          0  ...          0   

   day_of_week_fri  day_of_week_mon  day_of_week_thu  day_of_week_tue  \
0                0                1                0                0   
1                0                1                0                0   
2                0                1                0                0   
3                0                1                0                0   
4                0                1                0                0   

   day_of_week_wed  poutcome_failure  poutcome_nonexistent  poutcome_success  \
0                0                 0                     1                 0   
1                0                 0                     1                 0   
2                0                 0                     1                 0   
3                0                 0                     1                 0   
4                0                 0                     1                 0   

   y_yes  
0      0  
1      0  
2      0  
3      0  
4      0  

[5 rows x 51 columns]

In [32]:
bank_preprocess_one_hot.columns
X = bank_preprocess_one_hot.drop(columns=['y_yes'])
y = bank_preprocess_one_hot['y_yes']

In [33]:
ssc = StandardScaler(with_mean=True,with_std=True)
bank_preprocess_one_hot = ssc.fit_transform(X)

In [35]:
bank_preprocess_one_hot

array([[ 1.53307493,  0.02695841, -0.67334608, ..., -0.34102424,
         0.39974677, -0.18645954],
       [ 1.62899827, -0.44395837, -0.67334608, ..., -0.34102424,
         0.39974677, -0.18645954],
       [-0.28946853, -0.12020309, -0.67334608, ..., -0.34102424,
         0.39974677, -0.18645954],
       ...,
       [ 1.53307493, -0.27577381, -0.19880218, ..., -0.34102424,
         0.39974677, -0.18645954],
       [ 0.38199485,  0.78799356, -0.67334608, ..., -0.34102424,
         0.39974677, -0.18645954],
       [ 3.25969505, -0.0655431 ,  0.27574172, ...,  2.93234286,
        -2.50158369, -0.18645954]])

In [40]:
x_train, x_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=42)


C:\Users\krish\anacondanew\lib\site-packages\sklearn\linear_model\_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.9124463519313305

Logistic Regression

In [79]:
log = LogisticRegression()
log.fit(x_train,y_train)
y_pred = log.predict(x_test)
print('Accuracy Score : %f' % (accuracy_score(y_pred, y_test)))
parameters = {'C':[0.001, 0.1, 1, 10, 100]}
log_gsmodel = GridSearchCV(estimator=log, param_grid = parameters)
log_gsmodel_result = gsmodel.fit(x_train,y_train)
print("Best: %f using %s" % (log_gsmodel_result.best_score_, log_gsmodel_result.best_params_))
print('mean_test_score : %s' % (log_gsmodel_result.cv_results_['mean_test_score']))
print('std_test_Score : %s' % (log_gsmodel_result.cv_results_['std_test_score']))
print('params: %s' %(log_gsmodel_result.cv_results_['params']))

Accuracy Score : 0.912446
Best: 0.908676 using {'C': 100}
mean_test_score : [0.90499693 0.90803188 0.90545677 0.90803188 0.90867566]
std_test_Score : [0.00233751 0.00232865 0.00381591 0.00170134 0.00243209]
params: [{'C': 0.001}, {'C': 0.1}, {'C': 1}, {'C': 10}, {'C': 100}]


Decision Tree Classifier<br>
<br>
Entropy

In [53]:
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn import metrics

ds_tree_entropy = DecisionTreeClassifier(criterion='entropy',random_state=42)
ds_tree_entropy.fit(x_train,y_train)
ds_tree_entropy_y_pred = ds_tree_entropy.predict(x_test)
ds_tree_entropy_y_pred

array([0, 0, 0, ..., 0, 1, 0], dtype=uint8)

In [55]:
print("Model Entropy - no max depth")
print("Accuracy:", metrics.accuracy_score(y_test,ds_tree_entropy_y_pred))
print("Balanced accuracy:", metrics.balanced_accuracy_score(y_test,ds_tree_entropy_y_pred))
print('Precision score for "Yes"' , metrics.precision_score(y_test,ds_tree_entropy_y_pred, pos_label = 1))
print('Precision score for "No"' , metrics.precision_score(y_test,ds_tree_entropy_y_pred, pos_label = 0))
print('Recall score for "Yes"' , metrics.recall_score(y_test,ds_tree_entropy_y_pred, pos_label = 1))
print('Recall score for "No"' , metrics.recall_score(y_test,ds_tree_entropy_y_pred, pos_label = 0))

Model Entropy - no max depth
Accuracy: 0.8897608828939301
Balanced accuracy: 0.7267138225271417
Precision score for "Yes" 0.490343347639485
Precision score for "No" 0.9412986293783746
Recall score for "Yes" 0.5187287173666288
Recall score for "No" 0.9346989276876546


### Decision Tree - Entropy - GridSearch <br>

In [58]:
entr_parameters = {'max_depth':[2,3,4,5,6,7]}
dt_entr_gsmodel = GridSearchCV(estimator=ds_tree_entropy, param_grid = entr_parameters)
dt_entr_gsmodel_result = dt_entr_gsmodel.fit(x_train,y_train)
print("Best: %f using %s" % (dt_entr_gsmodel_result.best_score_, dt_entr_gsmodel_result.best_params_))
print('mean_test_score : %s' % (dt_entr_gsmodel_result.cv_results_['mean_test_score']))
print('std_test_Score : %s' % (dt_entr_gsmodel_result.cv_results_['std_test_score']))
print('params: %s' %(dt_entr_gsmodel_result.cv_results_['params']))

Best: 0.913611 using {'max_depth': 5}
mean_test_score : [0.90177805 0.90058246 0.9111588  0.91361128 0.91361128 0.91198651]
std_test_Score : [0.00158525 0.00175196 0.00190117 0.00103325 0.00147213 0.00099053]
params: [{'max_depth': 2}, {'max_depth': 3}, {'max_depth': 4}, {'max_depth': 5}, {'max_depth': 6}, {'max_depth': 7}]


Decision Tree<br><br>
Gini

In [56]:
ds_tree_gini = DecisionTreeClassifier(criterion='gini',random_state=42)
ds_tree_gini.fit(x_train,y_train)
ds_tree_gini_y_pred = ds_tree_gini.predict(x_test)
print("Model Entropy - no max depth")
print("Accuracy:", metrics.accuracy_score(y_test,ds_tree_gini_y_pred))
print("Balanced accuracy:", metrics.balanced_accuracy_score(y_test,ds_tree_gini_y_pred))
print('Precision score for "Yes"' , metrics.precision_score(y_test,ds_tree_gini_y_pred, pos_label = 1))
print('Precision score for "No"' , metrics.precision_score(y_test,ds_tree_gini_y_pred, pos_label = 0))
print('Recall score for "Yes"' , metrics.recall_score(y_test,ds_tree_gini_y_pred, pos_label = 1))
print('Recall score for "No"' , metrics.recall_score(y_test,ds_tree_gini_y_pred, pos_label = 0))

Model Entropy - no max depth
Accuracy: 0.8884120171673819
Balanced accuracy: 0.7254589059286929
Precision score for "Yes" 0.48459086078639746
Precision score for "No" 0.9410867757138897
Recall score for "Yes" 0.5175936435868331
Recall score for "No" 0.9333241682705526


### Decision Tree - Gini - GridSearch<br>

In [59]:
gini_parameters = {'max_depth':[2,3,4,5,6,7]}
dt_gini_gsmodel = GridSearchCV(estimator=ds_tree_gini, param_grid = gini_parameters)
dt_gini_gsmodel_result = dt_gini_gsmodel.fit(x_train,y_train)
print("Best: %f using %s" % (dt_gini_gsmodel_result.best_score_, dt_entr_gsmodel_result.best_params_))
print('mean_test_score : %s' % (dt_gini_gsmodel_result.cv_results_['mean_test_score']))
print('std_test_Score : %s' % (dt_gini_gsmodel_result.cv_results_['std_test_score']))
print('params: %s' %(dt_gini_gsmodel_result.cv_results_['params']))

Best: 0.913305 using {'max_depth': 5}
mean_test_score : [0.89969344 0.90576334 0.91112814 0.9131821  0.91330472 0.91213979]
std_test_Score : [0.00319559 0.00233952 0.00376758 0.00130999 0.00139679 0.00095182]
params: [{'max_depth': 2}, {'max_depth': 3}, {'max_depth': 4}, {'max_depth': 5}, {'max_depth': 6}, {'max_depth': 7}]


## Naive Baeys Classifier 

In [62]:
from sklearn.naive_bayes import GaussianNB
nb = GaussianNB()
nb_result = nb.fit(x_train,y_train)
nb_y_pred = nb.predict(x_test)
print(accuracy_score(nb_y_pred,y_test))

0.8681790312691601


## Stochastic Gradient Descent<br>

In [63]:
from sklearn.linear_model import SGDClassifier
sgd = SGDClassifier(loss='modified_huber',shuffle=True,random_state=42)
sgd.fit(x_train,y_train)
sgd_y_pred = sgd.predict(x_test)
print(accuracy_score(sgd_y_pred,y_test))

0.8925812385039853


## k-nearest neighbors

In [65]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(x_train,y_train)
knn_y_pred = knn.predict(x_test)
print(accuracy_score(knn_y_pred,y_test))

0.9011649294911097


In [68]:
knn_parameters = {'n_neighbors': range(10,20)}
knn_gsmodel = GridSearchCV(estimator=KNeighborsClassifier(), param_grid = knn_parameters)
knn_gsmodel_result = knn_gsmodel.fit(x_train,y_train)
print("Best: %f using %s" % (knn_gsmodel_result.best_score_, knn_gsmodel_result.best_params_))
print('mean_test_score : %s' % (knn_gsmodel_result.cv_results_['mean_test_score']))
print('std_test_Score : %s' % (knn_gsmodel_result.cv_results_['std_test_score']))
print('params: %s' %(knn_gsmodel_result.cv_results_['params']))

Best: 0.907020 using {'n_neighbors': 15}
mean_test_score : [0.89619865 0.8997241  0.90137952 0.90162477 0.90352544 0.90367872
 0.90487431 0.90564071 0.9060699  0.9063458  0.90674433 0.90619252
 0.90702023 0.90613121 0.90652974 0.90652974 0.90573268]
std_test_Score : [0.00105485 0.00107252 0.00034411 0.00255275 0.00238763 0.00255679
 0.00211193 0.00235194 0.00139274 0.00227144 0.00146893 0.00225692
 0.0016181  0.00103325 0.00124676 0.0014522  0.00180063]
params: [{'n_neighbors': 3}, {'n_neighbors': 4}, {'n_neighbors': 5}, {'n_neighbors': 6}, {'n_neighbors': 7}, {'n_neighbors': 8}, {'n_neighbors': 9}, {'n_neighbors': 10}, {'n_neighbors': 11}, {'n_neighbors': 12}, {'n_neighbors': 13}, {'n_neighbors': 14}, {'n_neighbors': 15}, {'n_neighbors': 16}, {'n_neighbors': 17}, {'n_neighbors': 18}, {'n_neighbors': 19}]


In [69]:
## Random Forest 

In [71]:
from sklearn.ensemble import RandomForestClassifier
rfm = RandomForestClassifier(n_estimators=20,oob_score=True,n_jobs=1,random_state=42,max_features=None,min_samples_leaf=10)
rfm.fit(x_train,y_train)
rfm_y_pred = rfm.predict(x_test)
print(accuracy_score(rfm_y_pred,y_test))

0.9144083384426732


In [76]:
rfm_parameters = {'n_estimators': [80,90,100]}
rfm_gsmodel = GridSearchCV(estimator=RandomForestClassifier(oob_score=True,n_jobs=1,random_state=42,max_features=None,min_samples_leaf=10), param_grid = rfm_parameters)
rfm_gsmodel_result = rfm_gsmodel.fit(x_train,y_train)
print("Best: %f using %s" % (rfm_gsmodel_result.best_score_, rfm_gsmodel_result.best_params_))
print('mean_test_score : %s' % (rfm_gsmodel_result.cv_results_['mean_test_score']))
print('std_test_Score : %s' % (rfm_gsmodel_result.cv_results_['std_test_score']))
print('params: %s' %(rfm_gsmodel_result.cv_results_['params']))

Best: 0.915297 using {'n_estimators': 90}
mean_test_score : [0.91459227 0.9147149  0.91486818 0.91489884 0.91529736 0.91517474]
std_test_Score : [0.00127656 0.00134887 0.00126621 0.00127656 0.00109935 0.0009125 ]
params: [{'n_estimators': 50}, {'n_estimators': 60}, {'n_estimators': 70}, {'n_estimators': 80}, {'n_estimators': 90}, {'n_estimators': 100}]


In [78]:
rfm_gsmodel_result.best_estimator_

RandomForestClassifier(max_features=None, min_samples_leaf=10, n_estimators=90,
                       n_jobs=1, oob_score=True, random_state=42)

In [80]:
from sklearn.ensemble import VotingClassifier
vote_classify = VotingClassifier(estimators=[
    ('log_be', log_gsmodel_result.best_estimator_), # Logistic Regression
    ('dt_entr_be', dt_entr_gsmodel_result.best_estimator_), #Decision tree entropy
    ('dt_gini_be', dt_gini_gsmodel_result.best_estimator_), #Decision tree Gini
    ('nb_be', nb), #Naive bayes
    ('sgd_be', sgd), #Stocastic Gradient Descent
    ('knn_be', knn_gsmodel_result.best_estimator_), #K-nearest Neighbors
    ('rfm_be', rfm_gsmodel_result.best_estimator_)], voting='hard') # Random Forest
vote_classify_model = vote_classify.fit(x_train, y_train)
vote_classify_ypred = vote_classify_model.predict(x_test)
print(accuracy_score(vote_classify_ypred,y_test))

0.9174739423666463


In [81]:
from sklearn.metrics import confusion_matrix
print(confusion_matrix(vote_classify_ypred,y_test))

[[7058  457]
 [ 216  424]]


In [83]:
y.value_counts()

0    36198
1     4577
Name: y_yes, dtype: int64

In [84]:
4577/36198

0.1264434499143599

In [85]:
y_test.value_counts()

0    7274
1     881
Name: y_yes, dtype: int64

In [86]:
881/7274

0.12111630464668682

7058(correctly predicted 0) + 216(Falsely predicted 0) = 7274 (total 0) <br>
424(correctly predicted 1) + 457(Falsely predicted 1) = 881 (total 1)<b>
we are able to predict almost half of positive deposits! and predict fairly well who wont deposit. This will help navigate between the groups for better turn over